smart-contracts/contracts/Unlock.sol

```diff
@@ -509,6 +528,112 @@ contract Unlock is UnlockInitializable, UnlockOwnable {
     }
   }
 
+  function getBalance(address token) internal view returns (uint) {
+    return token == address(0) ?
+      address(this).balance 
+      :
+      IERC20(token).balanceOf(address(this));
+  }
+
+  /**
+   * Swap tokens and call a function in a lock contract
+   * @notice If the actual amount spent is less than the specified maximum amount, the remaining tokens wll 
+   * be held by the Unlock contract
+   */
+  function swapAndCall(
+    address lock,
+    address srcToken,
+    uint amountInMax,
+    address swapRouter,
+    bytes memory swapCalldata,
+    bytes memory callData
+  ) public payable returns(bytes memory) {
+    // check if lock exists
+    if(!locks[lock].deployed) {
+      revert('l');
+    }
+
+    // get lock pricing 
+    address destToken = IPublicLock(lock).tokenAddress();
+
+    // get price in destToken from lock
+    uint keyPrice = IPublicLock(lock).keyPrice();
+
+    // get balances of Unlock before
+    // if payments in ETH, substract the value sent by user to get actual balance
```

Comment:
Hum, we should not hae received any `destTokens` though... msg.value will in srcToken,